### PR TITLE
[4.0] convert to prepared statements User Table

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
converted sql to prepared statements

### Testing Instructions

- create new users
- delete users
- modify users
- register

### Expected result
works as before

### Actual result
N/A

### CODE REVIEW

File: libraries/src/Table/User.php [hunk 1]
```diff
@@ -372,16 +391,21 @@ public function store($updateNulls = true)
...
-							-> where($this->_db->quoteName('user_id') . ' = ' . (int) $this->id)
-							-> where($this->_db->quoteName('group_id') . ' = ' . (int) $map->group_id);
-
-						$this->_db->setQuery($query);
-						$this->_db->execute();
+						$mapGroupId[] = (int) $map->group_id;
 					}
 				}
+
+				if (\count($mapGroupId))
```
Comment (wilsonge) [inline libraries/src/Table/User.php:400]: I think you need to do an initialise call for this variable? I don't think count checks if the variable is defined

**Review by wilsonge — APPROVED** [review]

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on d1924e9b21dbb05ba551e3508096fbd5bd28709d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28479">issues.joomla.org/tracker/joomla-cms/28479</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on d1924e9b21dbb05ba551e3508096fbd5bd28709d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28479">issues.joomla.org/tracker/joomla-cms/28479</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28479">issues.joomla.org/tracker/joomla-cms/28479</a>.</sub>

**Comment by Quy** [comment]
> PHP Notice:  Undefined variable: mapGroupId in \libraries\src\Table\User.php on line 398

> PHP Warning:  count(): Parameter must be an array or an object that implements Countable in \libraries\src\Table\User.php on line 398

File: libraries/src/Table/User.php (hunk 1 above)
Comment (alikon) [reply libraries/src/Table/User.php:400]: done 
ths

**Comment by alikon** [comment]
notice/warning fixed

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8268a322f32e8794264f49641b12b40fb51df614<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28479">issues.joomla.org/tracker/joomla-cms/28479</a>.</sub>

File: libraries/src/Table/User.php [hunk 2]
```diff
@@ -502,14 +543,18 @@ public function setLastVisit($timeStamp = null, $userId = null)
...
+		$lastVisit = $date->toSql();
 
 		// Update the database row for the user.
 		$db = $this->_db;
 		$query = $db->getQuery(true)
 			->update($db->quoteName($this->_tbl))
-			->set($db->quoteName('lastvisitDate') . '=' . $db->quote($date->toSql()))
-			->where($db->quoteName('id') . '=' . (int) $userId);
+			->set($db->quoteName('lastvisitDate') . ' = :lastvisitDate')
+			->where($db->quoteName('id') . '= :id')
```
Comment (Quy) [inline libraries/src/Table/User.php]:
```suggestion
			->where($db->quoteName('id') . ' = :id')
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8268a322f32e8794264f49641b12b40fb51df614<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28479">issues.joomla.org/tracker/joomla-cms/28479</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**